zkvm/docs/zkvm-blockchain.md

```diff
@@ -0,0 +1,386 @@
+   - `utxos`: empty set
+   - `nonces`: empty set
+   - `refids`: empty set
+
+## Make initial block
+
+Inputs:
+- `timestamp_ms`,
+  the current time as a number of milliseconds since the Unix epoch:
+  00:00:00 UTC Jan 1,
+  1970.
+- `refscount`,
+  the number of recent block ids to cache for
+  [nonce](zkvm-spec.md#nonce)
+  uniqueness.
+
+Output:
+- A [block](#block).
+
+Procedure:
+1. [Compute txroot](#compute-txroot) from an empty list of transactions.
+2. [Compute utxoroot](#compute-utxoroot) from an empty set of utxos.
+3. [Compute nonceroot](#compute-nonceroot) from an empty set of nonce anchors.
+4. Compute `header`,
+   a
+   [block header](#block-header)
+   with its fields set as follows:
+   - `version`: 1
+   - `height`: 1
+   - `previd`: all-zero string of 32-bytes
+   - `timestamp_ms`: `timestamp_ms`
+   - `txroot`: `txroot`
+   - `utxoroot`: `utxoroot`
+   - `nonceroot`: `nonceroot`
+   - `refscount`: `refscount`
+   - `ext`: empty
+5. Return a block with `header` set to `header` and an empty `txs` list.
+
+## Join existing network
+
+A new node starts here when joining a running network.
+It must either:
+- obtain all historical blocks, [applying](#apply-block) them one by one to reproduce the latest [blockchain state](#blockchain-state); or
+- obtain a recent copy of the blockchain state `state` from a trusted source
+  (e.g., another node that has already validated the full history of the blockchain)
+  and begin applying blocks beginning at `state.tipheader.height+1`.
+
+An obtained (as opposed to computed) blockchain state `state` may be partially validated by [computing the utxoroot](#compute-utxoroot) from `state.utxos` and verifying that it equals `state.header.utxoroot`.
+
+
+## Validate block
+
+Validating a block checks it for correctness outside the context of a particular [blockchain state](#blockchain-state).
+
+Additional correctness checks against a particular blockchain state happen during the [apply block](#apply-block) procedure.
+
+Inputs:
+- `block`,
+  the block to validate.
+- `prevheader`,
+  the header of the previous block.
+
+Output:
+- list of [transaction logs](zkvm-spec.md#transaction-log),
+  one for each transaction in block.txs.
+
+Procedure:
+1. Verify `block.header.version >= prevheader.version`.
+2. If `block.header.version == 1`, verify `block.header.ext` is empty.
+3. Verify `block.header.height == prevheader.height+1`.
+4. Verify `block.header.previd` equals the
+   [block ID](#block-id)
+   of `prevheader`.
+5. Verify `block.header.timestamp_ms > prevheader.timestamp_ms`.
+6. Verify `block.header.refscount >= 0` and `block.header.refscount <= prevheader.refscount + 1`.
+7. [Compute txroot](#compute-txroot) from `block.txs`.
+8. Verify `txroot == block.header.txroot`.
+9. For each transaction `tx` in block.txs:
+   1. Verify `tx.mintime_ms <= block.header.timestamp_ms <= tx.maxtime_ms`.
+   2. If `block.header.version == 1`,
+      verify `tx.version == 1`.
```

Comment:
Also verify that `ext` is empty to enforce that it's not abused until there's a version upgrade.

Reply:
See step 2 above.

Reply:
👍